### ISSUE TITLE
Cannot access host-bound ports when running a service with multiple networks

### ISSUE THREAD

**Issue by ubi-US** [docker/compose#3055]
The new virtual networks in docker 1.10 and compose 1.6 are exciting. Yet, I stumbled across the following issue when exposing a port to the outside world:

I have a composition of multiple services, where these services are attached to various networks. One of the services exposes a port to the outside world, so that the web service it offers can be access from other hosts. Let's say service A exposes a port, via the `ports` keyword, say `"45000:80"`; that is, the port 45000 should be accessible from any client.

My setup is running on a single host, with the bridge network driver only (no overlay networks, no multi-host networks).
- In the simplest setting, , both services A and B are on the same network. Then, the exposed port 45000 is accessible both from `localhost` as well as over the internet.
- If I add another network, and have service A connect to one network and service B to another network, the exposed port 45000 of service A is still reachable both from `localhost` as well as over the internet.
- However, if I connect service A to both networks, its exposed port 45000 is no longer accessible over the internet, but it is still accessible from `localhost`. If I expose another port on service B - the service that is connected to a single network only - then this port is accessible from the internet.

I am no networking expert, so here are my questions
- Is this a bug or intended behavior?
- Do i need to adjust my iptables in case I want to make the port on the multi-network service available externally?

Here is my debugging info
I am using docker machine to connect to a ubuntu box using the generic driver
_docker info_

```
Server Version: 1.10.2
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 4.2.0-30-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 19.61 GiB
Name: corehost
ID: TD2T:D3R2:E6HT:GJYS:GEIH:FYM3:VUZC:755D:AB7X:RR7Q:UFNQ:CK3J
WARNING: No swap limit support
Labels:
 provider=generic
```

_docker version_

```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64
```

_docker-compose version_

```
docker-compose version 1.6.2, build 4d72027
docker-py version: 1.7.2
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013

```

**Comment by dnephin** [comment]
cc @mavenugo

I'm not sure honestly. It seems strange that connecting to a second network would be an issue. I would think it always uses the same host interface.

**Comment by mavenugo** [comment]
@ubi-US @dnephin it seems like a bug to me. we will take a look at it.

**Comment by linfan** [comment]
Same issue here, also I find other containers in the same overlay network are not added into the container's /etc/hosts file, so docker-compose configure file which works for single host is not working in Swarm any more.

**Comment by dnephin** [comment]
>  I find other containers in the same overlay network are not added into the container's /etc/hosts

This is expected, embeded DNS is used now, not `/etc/hosts`. Not related to this issue.

**Comment by darkermatter** [comment]
I'm observing the same thing, but it gets stranger...

docker-compose.yml snippet:
```
    ports:
      - "25:25"
    networks:
      - core
      - default
```
iptables-save:
```
# Generated by iptables-save v1.4.21 on Tue Mar  8 21:07:23 2016
*nat
:PREROUTING ACCEPT [2716:412115]
:INPUT ACCEPT [2712:411700]
:OUTPUT ACCEPT [377:26148]
:POSTROUTING ACCEPT [381:26563]
:DOCKER - [0:0]
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT ! -d 127.0.0.0/8 -m addrtype --dst-type LOCAL -j DOCKER
-A POSTROUTING -s 172.18.0.0/16 ! -o br-02738ab0a2e2 -j MASQUERADE
-A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE
-A POSTROUTING -s 172.18.0.5/32 -d 172.18.0.5/32 -p tcp -m tcp --dport 443 -j MASQUERADE
-A POSTROUTING -s 172.18.0.5/32 -d 172.18.0.5/32 -p tcp -m tcp --dport 80 -j MASQUERADE
-A DOCKER -i br-02738ab0a2e2 -j RETURN
-A DOCKER -i docker0 -j RETURN
-A DOCKER ! -i br-02738ab0a2e2 -p tcp -m tcp --dport 443 -j DNAT --to-destination 172.18.0.5:443
-A DOCKER ! -i br-02738ab0a2e2 -p tcp -m tcp --dport 80 -j DNAT --to-destination 172.18.0.5:80
COMMIT
# Completed on Tue Mar  8 21:07:23 2016
# Generated by iptables-save v1.4.21 on Tue Mar  8 21:07:23 2016
*filter
:INPUT ACCEPT [139987:12409715]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [208879:190554919]
:DOCKER - [0:0]
:DOCKER-ISOLATION - [0:0]
-A FORWARD -j DOCKER-ISOLATION
-A FORWARD -o br-02738ab0a2e2 -j DOCKER
-A FORWARD -o br-02738ab0a2e2 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i br-02738ab0a2e2 ! -o br-02738ab0a2e2 -j ACCEPT
-A FORWARD -i br-02738ab0a2e2 -o br-02738ab0a2e2 -j ACCEPT
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
-A DOCKER -d 172.18.0.5/32 ! -i br-02738ab0a2e2 -o br-02738ab0a2e2 -p tcp -m tcp --dport 443 -j ACCEPT
-A DOCKER -d 172.18.0.5/32 ! -i br-02738ab0a2e2 -o br-02738ab0a2e2 -p tcp -m tcp --dport 80 -j ACCEPT
-A DOCKER-ISOLATION -i docker0 -o br-02738ab0a2e2 -j DROP
-A DOCKER-ISOLATION -i br-02738ab0a2e2 -o docker0 -j DROP
-A DOCKER-ISOLATION -j RETURN
COMMIT
# Completed on Tue Mar  8 21:07:23 2016
```
netstat:
```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:2049            0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:52001           0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:52360           0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:111             0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:22000           0.0.0.0:*               LISTEN      14/sshd
tcp        0      0 0.0.0.0:44179           0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:36316           0.0.0.0:*               LISTEN      -
tcp6       0      0 :::2049                 :::*                    LISTEN      -
tcp6       0      0 :::2376                 :::*                    LISTEN      -
tcp6       0      0 :::111                  :::*                    LISTEN      -
tcp6       0      0 :::80                   :::*                    LISTEN      -
tcp6       0      0 :::22000                :::*                    LISTEN      14/sshd
tcp6       0      0 :::50768                :::*                    LISTEN      -
tcp6       0      0 :::46482                :::*                    LISTEN      -
tcp6       0      0 :::22                   :::*                    LISTEN      -
tcp6       0      0 :::443                  :::*                    LISTEN      -
tcp6       0      0 :::49980                :::*                    LISTEN      -
tcp6       0      0 :::39869                :::*                    LISTEN      -
```
and now for the strange bit...
```
$ telnet 192.168.144.4 25
Trying 192.168.144.4...
Connected to 192.168.144.4.
Escape character is '^]'.
220 rt.serviomatic.com ESMTP (Debian/GNU)
quit
221 2.0.0 Bye
Connection closed by foreign host.
```
and networking inside the container:
```
root@rt:/# ifconfig
eth0      Link encap:Ethernet  HWaddr 02:42:ac:13:00:02
          inet addr:172.19.0.2  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe13:2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:106082 errors:0 dropped:0 overruns:0 frame:0
          TX packets:98875 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:15340761 (14.6 MiB)  TX bytes:235001078 (224.1 MiB)

eth1      Link encap:Ethernet  HWaddr 02:42:ac:15:00:03
          inet addr:172.21.0.3  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe15:3/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:193872 errors:0 dropped:0 overruns:0 frame:0
          TX packets:264879 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:426625058 (406.8 MiB)  TX bytes:38116016 (36.3 MiB)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:746 errors:0 dropped:0 overruns:0 frame:0
          TX packets:746 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:539367 (526.7 KiB)  TX bytes:539367 (526.7 KiB)

root@rt:/# netstat -tulnp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      113/nginx
tcp        0      0 0.0.0.0:25              0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.11:41147        0.0.0.0:*               LISTEN      -
tcp6       0      0 :::25                   :::*                    LISTEN      -
udp        0      0 127.0.0.11:43614        0.0.0.0:*                           -
root@rt:/# exit
```

I can't see how port 25 is getting through!

```
Client:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 19:55:25 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 18:41:30 2016
 OS/Arch:      linux/amd64
```

**Comment by darkermatter** [comment]
Oops, never mind the strange bit - I didn't notice I was inside another container when I ran the netstat, etc. I can see the docker-proxy process on the host for port 25 - how embarrassing. A real Inception moment! :-)

However, I have another container (gitlab) with the same networking setup which should be listening on port 10022 and isn't:

```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:11942           0.0.0.0:*               LISTEN      1505/openvpn
tcp        0      0 0.0.0.0:110             0.0.0.0:*               LISTEN      1638/openvpn
tcp        0      0 0.0.0.0:22000           0.0.0.0:*               LISTEN      21404/sshd
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      665/sshd
tcp6       0      0 :::6880                 :::*                    LISTEN      1459/docker-proxy
tcp6       0      0 :::10051                :::*                    LISTEN      1839/docker-proxy
tcp6       0      0 :::10022                :::*                    LISTEN      47021/docker-proxy
tcp6       0      0 :::2376                 :::*                    LISTEN      713/docker
tcp6       0      0 :::22000                :::*                    LISTEN      21404/sshd
tcp6       0      0 :::80                   :::*                    LISTEN      1728/docker-proxy
tcp6       0      0 :::22                   :::*                    LISTEN      665/sshd
tcp6       0      0 :::25                   :::*                    LISTEN      33355/docker-proxy
tcp6       0      0 :::443                  :::*                    LISTEN      1719/docker-proxy
udp        0      0 0.0.0.0:34744           0.0.0.0:*                           1092/dhclient
udp        0      0 0.0.0.0:11941           0.0.0.0:*                           1588/openvpn
udp        0      0 0.0.0.0:11942           0.0.0.0:*                           1618/openvpn
udp        0      0 0.0.0.0:11943           0.0.0.0:*                           2087/openvpn
udp        0      0 0.0.0.0:68              0.0.0.0:*                           1092/dhclient
udp6       0      0 :::40904                :::*                                1092/dhclient
```

**Comment by darkermatter** [comment]
Oops, I didn't notice I was inside another container when I did the netstat - I can see the docker-proxy processes running directly on the host. How embarrassing! A real Inception moment.

**Comment by darkermatter** [comment]
I did some more testing...

I was getting timeouts trying to push to gitlab while the port forwarding was bound to eth1:
```
eth0      Link encap:Ethernet  HWaddr 02:42:ac:13:00:06
          inet addr:172.19.0.6  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe13:6/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:12 errors:0 dropped:0 overruns:0 frame:0
          TX packets:12 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:928 (928.0 B)  TX bytes:928 (928.0 B)

eth1      Link encap:Ethernet  HWaddr 02:42:ac:14:00:05
          inet addr:172.20.0.5  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe14:5/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:873 errors:0 dropped:0 overruns:0 frame:0
          TX packets:659 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:139650 (139.6 KB)  TX bytes:199297 (199.2 KB)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:134 errors:0 dropped:0 overruns:0 frame:0
          TX packets:134 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:10098 (10.0 KB)  TX bytes:10098 (10.0 KB)
```
snippets from iptables-save:
```
-A DOCKER ! -i br-f1963208cc56 -p tcp -m tcp --dport 10022 -j DNAT --to-destination 172.20.0.5:22
-A DOCKER -d 172.20.0.5/32 ! -i br-f1963208cc56 -o br-f1963208cc56 -p tcp -m tcp --dport 22 -j ACCEPT
```
Full output: https://gist.github.com/darkermatter/0c24ffb658d86383ba12

After restarting the container:
```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.11:46206        0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.1:8080          0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      476/nginx -g daemon
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      477/sshd
tcp6       0      0 :::80                   :::*                    LISTEN      476/nginx -g daemon
tcp6       0      0 :::22                   :::*                    LISTEN      477/sshd
udp        0      0 127.0.0.11:58311        0.0.0.0:*                           -
```
```
-A DOCKER ! -i br-8aaf89de458f -p tcp -m tcp --dport 10022 -j DNAT --to-destination 172.19.0.6:22
-A DOCKER -d 172.19.0.6/32 ! -i br-8aaf89de458f -o br-8aaf89de458f -p tcp -m tcp --dport 22 -j ACCEPT
```
Full output: https://gist.github.com/darkermatter/5579257b8b74b981325e

The port is being forwarded through to eth0 and I can now push to gitlab. It looks like ssh is listening on all interfaces so that shouldn't be an issue. Perhaps there's a clue in the full output of iptables-save but I've been working for 16 hours and maybe I'm missing something obvious.

**Comment by mavenugo** [comment]
This issue should be fixed in docker master now. Can someone confirm that ?

**Comment by rysiekpl** [comment]
@mavenugo link to `docker` master? I am experiencing this issue on `docker` 1.10.3, `docker-compose` 1.6.2 when a container is member of more than 3 networks. [Testcase available](https://git.occrp.org/libre/docker-netfail).

**Comment by R-omk** [comment]
I have same problem.
The fact is that a container have default gateway from first network eth0, but  host build wrong iptables rule:

 `# iptables -L -n -t nat`

```
<...>
RETURN     all  --  0.0.0.0/0            0.0.0.0/0           
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:10080 to:172.21.0.2:3000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:10022 to:172.21.0.2:22

```

when container's gateway is 172.23.0.1 


So 

```
    networks:
      - innernet   # 172.23.0.2/16    eth0 
      - externalnet  # 172.21.0.2/16  eth1
```

Is it possible to make the order in the list of networks affect the network settings inside the container?


Any workaround ideas to get it works properly?

**Comment by railnet** [comment]
It seems an active discussion on similar symptoms is active here too:
https://github.com/docker/docker/issues/21741
@ubi-US @dnephin @mavenugo @linfan @darkermatter @bmerry @aboch @JosephEarl @oben59 @choobs-dev

May you confirm I'm right?
I'm available for possible further investigation in order to support you in the fix.

**Comment by dustymabe** [comment]
This is my attempt at an explanation of this problem:
http://dustymabe.com/2016/05/25/non-deterministic-docker-networking-and-source-based-ip-routing/

**Comment by stouf** [comment]
I don't know what is the status of this issue, but it has been now 2 days that I'm fighting with this bug, trying to understand all the small bits of it before deploying an update to my production environment (by the way, @dustymabe , thanks for the article you gave a link to, it made the things really easy to understand).
After these 2 crazy days, I finally gave up on getting this to work on Docker **v1.10.3** and updated Docker to the version **1.11.2**. Now, it works like a charm. I even tried to break it on purpose by disconnecting a container from the network where its default gateway was set to, but Docker seems to dynamically reconfigure the iptables rules and the container network settings so that nothing breaks 👍

**Comment by shin-** [comment]
Fixed in Docker 1.11
